Take part document

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Handle Take Part documents in Frontend, in preparation for them being taken out of Government Frontend

## Why

https://trello.com/c/0LYU5bSd/318-move-route-document-type-takepart-from-governmentfrontend-to-frontend, [Jira issue PNP-8559](https://gov-uk.atlassian.net/browse/PNP-8559)

## How

These documents all have a fixed prefix, so add a scoped route for them, a controller, a model and a view. Here we add the initial ContentItem model the document type models will be based on, with the minimum functionality required to get Take Part documents working.

We also port across the figure component, which is used in this document type.

We also create a couple of helpers for rendering generic content.

## Screenshots?

From example page: https://www.gov.uk/government/get-involved/take-part/improve-your-social-housing

### Before (running in Government Frontend)

![www gov uk_government_get-involved_take-part_improve-your-social-housing (1)](https://github.com/user-attachments/assets/03278975-6d66-4b50-8ac1-811d990c860f)

### After (running in Frontend)

![frontend dev gov uk_government_get-involved_take-part_improve-your-social-housing (1)](https://github.com/user-attachments/assets/75149b10-0918-473f-8d82-d3e9fb506953)

KludgeKML committed:
Add ordered_related_items support

- Add code in ContentItem links into the desired internal format
  for the sidebar. Take part is the first document that uses
  ordered_related_links.
- Add tests

Commit audit trail:
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/controllers/content_items_controller.rb#L151-L157

leenagupte commented on config/locales/cy.yml:
```diff
@@ -102,6 +102,15 @@ cy:
       image_credit: 'Delwedd gan: %{credit}'
+  content_item:
+    schema_name:
+      take_part:
```
There's also a `formats` lower this file that contains content for the original frontend document types.

I think they should all be in the same place, wherever that is.

KludgeKML replied:
Moved to formats

KludgeKML committed:
Add take_part examples to dev controller and README.md

leenagupte commented on spec/system/take_part_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    content_store_has_example_item("/government/get-involved/take-part/tp1", schema: :take_part)
+  end
+
+  context "/government/get-involved/take-part/:slug" do
```
This is interesting. Should the `context` for system tests be set to the path? Is that needed because it's specified in the `visit` call. 
For other [document types](https://github.com/alphagov/frontend/blob/main/spec/system/place_spec.rb#L69) the context is more descriptive.

Not a blocker, but something to consider as it would be better if things were consistent.

KludgeKML replied:
Context changed to have a better name

leenagupte commented on spec/helpers/locale_helper_spec.rb:
```diff
@@ -0,0 +1,46 @@
+RSpec.describe LocaleHelper do
```
Nitpick: This should be added in the same commit as `LocaleHelper`. Perhaps they can be added together in their own commit? It would make the audit trail easier to follow.

KludgeKML replied:
Now added in same commit.

leenagupte commented on spec/helpers/locale_helper_spec.rb:
```diff
@@ -0,0 +1,46 @@
+
+  describe "#t_locale_fallback" do
+    it "returns nil if a value of the key is present" do
+      expect(t_locale_fallback("formats.take_part", count: 1)).to be_nil
```
So `t_locale_fallback` method returns the "fallback", i.e. the English translation, if the language key doesn't exist? I feel like the language keys _should_ exist, and they default to English by default if there's no value.

Actually, I've just looked at the [commit message](https://github.com/alphagov/government-frontend/commit/c409f0f01404b7ccb1792f330ce415aa11314bec) from when the original method was added and it's existence makes sense.

KludgeKML replied:
(Positive, so marking as resolved)

leenagupte commented on spec/models/content_item_spec.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+  end
+
+  describe "#available_translations" do
```
Ah! The model tests. These should be fixed up to where the `available_translations` method is added.

KludgeKML replied:
(Now in correct commit)

leenagupte commented on spec/system/meta_tags_spec.rb:
```diff
@@ -0,0 +1,45 @@
+RSpec.describe "Meta Tags" do
```
Well done for spotting these tests and bringing them over! 🎉 

I think it would really useful to have these tests as a shared example as all of the document types will need to render meta tags and they're an easy thing to miss. 

Perhaps convert them to shared examples where the content item and route being tested can be passed in?

KludgeKML replied:
Now shared.

leenagupte reviewed (commented):
This is looking really good! 🎉 

I have some inline comments. I think the biggest area of contention is locales and whether all 66 should be added. I think if there are document types that have multiple translations or could have them, then all of the language keys should be added.

KludgeKML commented on app/helpers/machine_readable_metadata_helper.rb:
```diff
@@ -0,0 +1,6 @@
+module MachineReadableMetadataHelper
+  def machine_readable_metadata(content_item, args)
```
Have got rid of this method, and made a shared example for testing the meta tags.

leenagupte commented on app/models/content_item.rb:
```diff
@@ -10,8 +10,20 @@ def initialize(content_store_response)
     @base_path = content_store_response["base_path"]
     @locale = content_store_response["locale"]
+
+    content_store_response["links"]["ordered_related_items"] = ordered_related_items(content_store_response["links"]) if content_store_response["links"]
```
Ah, it looks like `ordered_related_items` are being used, but in a slightly hidden way. The entire content_item is passed to the `contextual_sidebar` and `contextual_footer` components, and a [helper](https://github.com/alphagov/govuk_publishing_components/blob/main/lib/govuk_publishing_components/presenters/related_navigation_helper.rb#L122) used by those components figures out how to render any and all things in the links hash.

So to answer my own question, `ordered_related_items` are being used and this is needed.

KludgeKML committed:
Add Figure component

- Used in Take Part, among others
- Includes spec file, documentation

Commit audit trail:
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/assets/stylesheets/components/_figure.scss
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/views/components/_figure.html.erb
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/views/components/docs/figure.yml
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/test/components/figure_test.rb

KludgeKML committed:
Add translations for figure component

Commit audit trail:
- https://github.com/alphagov/government-frontend/tree/6523741691ddfb9551967311a3f79775e9403ecf/config/locales

KludgeKML commented on app/helpers/locale_helper.rb:
```diff
@@ -2,4 +2,34 @@ module LocaleHelper
+    I18n.t("language_names.#{locale}", locale:)
+  end
+
+  def translations_for_nav(translations)
```
Closing as positive.

KludgeKML committed:
Add locale keys for translation shared partial

Commit audit trail:
- https://github.com/alphagov/government-frontend/tree/6523741691ddfb9551967311a3f79775e9403ecf/config/locales

KludgeKML committed:
Add translation shared partial

- Add methods to LocaleHelper to provide language names and translate content info into component format
- Add methods to ContentItem to provide available translations

Commit audit trail:
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/helpers/application_helper.rb
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/views/shared/_translations.html.erb
- https://github.com/alphagov/government-frontend/blob/a938e4bee78b794e4e5303a1bde1dfd05e94b287/app/presenters/content_item_presenter.rb#L152-L154

KludgeKML committed:
Add TakePartController

- add view and helpers
- convert test suite to RSpec on move
- note that we explicitly remove the content-bottom-margin div
  class that was immediately inside the 2/3rds column. This is
  a style used in a handful of government-frontend document
  types which (on review with frontend developers) does not
  really fix the problem it's trying to fix (insufficient bottom
  margin on some tablet-sized devices). This has been captured
  as https://github.com/alphagov/govuk_publishing_components/issues/4220
  to be fixed properly in the components gem.
- We replace the call to TitleHelper in the view with a direct call,
  as it didn't add much and it's not clear whether it's worth
  copying that helper across yet.
- main element wrapper added as it's in the layout in government-
  frontend, but not in frontend.

Commit audit trail:
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/app/views/content_items/take_part.html.erb
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/test/integration/take_part_test.rb

KludgeKML committed:
Update consolidation task to be able to rename root key

KludgeKML committed:
Add take_part translation keys

- copied from government-frontend, originally this was in content_item.schema_name.take_part,
  here we put it in formats.take_part

Commit audit trail:
- https://github.com/alphagov/government-frontend/tree/6523741691ddfb9551967311a3f79775e9403ecf/config/locales

KludgeKML commented on app/models/content_item.rb:
```diff
@@ -17,6 +17,12 @@ def initialize(content_store_response)
   delegate :to_h, to: :content_store_response
   delegate :cache_control, to: :content_store_response
 
+  def available_translations
```
(Now in this commit)

KludgeKML committed:
Set default slug in ContentItemController to request.path

- In practise, it's almost always request.path, and the specific
  content items that need it to be something different are already
  overriding it, so better to configure it as default here

KludgeKML committed:
Shared tests for meta_tags with images

- Needed for the tests added in the next commit.
- Separate from the no-image tags test already here,
  as the shared examples for images break if run on a
  non-image-bearing document type like help pages

Commit audit trail:
- https://github.com/alphagov/government-frontend/blob/6523741691ddfb9551967311a3f79775e9403ecf/test/integration/meta_tags_test.rb

KludgeKML committed:
Simplify session_spec test

KludgeKML commented on config/locales/en.yml:
```diff
@@ -100,6 +100,9 @@ en:
     upcoming_bank_holidays: Upcoming bank holidays
+  components:
+    figure:
+      image_credit: 'Image credit: %{credit}'
```
All languages added (now in commit following the one the figure is added in)

KludgeKML commented on config/locales/en.yml:
```diff
@@ -650,6 +650,11 @@ en:
       - title: VAT rates
         link: /vat-rates
+  i18n:
+    direction: ltr
```
Change is now obsolete (handled in a previous PR)

KludgeKML commented on app/controllers/take_part_controller.rb:
```diff
@@ -0,0 +1,5 @@
+class TakePartController < ContentItemsController
+  def content_item_slug
+    request.path
```
Done

KludgeKML commented on app/helpers/locale_helper.rb:
```diff
@@ -2,4 +2,34 @@ module LocaleHelper
+  end
+
+  def page_text_direction
+    I18n.t("i18n.direction", locale: I18n.locale, default: "ltr")
```
Change included in a previous PR now.

KludgeKML commented on app/views/take_part/show.html.erb:
```diff
@@ -0,0 +1,37 @@
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= render 'govuk_publishing_components/components/title', title_and_context(@content_item.title, @content_item.document_type).except(:average_title_length) %>
```
Removed

KludgeKML commented on config/locales/en.yml:
```diff
@@ -447,6 +447,9 @@ en:
     start_now: Start now
+    take_part:
+      one: Take part
+      other: Take part
```
They vary by language, so some have one/other, some have few/many/etc. It's all handled by i18n

leenagupte commented on spec/support/meta_tags.rb:
```diff
@@ -18,3 +18,28 @@
+RSpec.shared_examples "it has meta tags for images" do |schema, base_path|
+  before do
+    example_doc = GovukSchemas::RandomExample.for_schema(frontend_schema: schema) do |random|
+      random["details"] = random["details"].merge(
```
Minor: I think you can do `random["details"].merge!(...)` to change the original object rather than an assign.

leenagupte commented on spec/system/take_part_spec.rb:
```diff
@@ -0,0 +1,27 @@
+RSpec.describe "TakePart" do
+  it_behaves_like "it has meta tags", "take_part", "/government/get-involved/take-part"
+  it_behaves_like "it has meta tags for images", "take_part", "/government/get-involved/take-part"
```
Love this!

leenagupte reviewed (commented):
On the whole the code looks good. 🎉 

However, the request tests are missing and should be added.

Also, I think every commit message should be checked to make sure that it's linking to the permalink to the file in government-frontend rather than the commit containing the latest change, as some of those contain a lot of files.

In some cases the file is being linked to, and even better, the actual lines being pulled across, but in those cases it would be better if it linked to the "code" rather than the "blame".

Once these things are fixed this PR will be good to go.

KludgeKML commented on app/assets/stylesheets/components/_figure.scss:
Done

KludgeKML commented on app/helpers/locale_helper.rb:
Done

KludgeKML commented on config/locales/ar.yml:
```diff
@@ -12,3 +12,70 @@ ar:
   i18n:
     direction: rtl
+  language_names:
+    ar: العربيَّة
```
Done

KludgeKML commented on spec/models/content_item_spec.rb:
```diff
@@ -1,6 +1,56 @@
 RSpec.describe ContentItem do
   let(:subject) { build(:content_item_with_data_attachments, schema_name: "fancy_page_type") }
 
+  describe "#initialize" do
```
I've called it `describe "ordered_related_items attribute"`

leenagupte reviewed (commented):
It's looking great. There's only one minor CSS file fix, mentioned inline, and a commit message that needs to be edited that we've chatted about elsewhere. Otherwise it looks good. 🎉

leenagupte reviewed (approved):
Looks great! Thanks for adding such a great audit trail of the files back to government-frontend. 🥇

KludgeKML committed:
Update wrapper to include direction

- We also update one test that is a little too specific
  about what the wrapper class should be.
